### PR TITLE
18: Implement relabeling of the replay buffer

### PR DESCRIPTION
**Pull request by jualat**
**Issue:** #18 

**What?**
- implement relabeling of the replay buffer

**How?**
- define `predict_reward` in `RewardNet`
- use it in `relabel_rewards` in `ReplayBuffer`

**Ready to merge to main?**
- [ ] Linter passes
- [ ] PR Approved

**Performance**
image

### CODE REVIEW

File: poetry.lock
Comment (bkuen) [inline poetry.lock:1]: As far as I know, we haven't pushed poetry.lock so far and I would keep it like this, at least until all the other necessary PRs are merged. Maybe then we could think about generating the lock file once with all necessary dependencies and update the main branch. But maybe it would also be okay to just keep it as it is.

**Review by bkuen — CHANGES_REQUESTED** [review]
The implementation looks good to me. Just remove the .lock file. Maybe someone else could take a look as well :)

**Comment by Wizzardexe** [comment]
Looks good, but we need to adjust where the groundTruthRewards are stored. With this, the rewards are updated in the buffer and used for comparison in the teacher. However, for accurate comparisons in the teacher, we should still use groundTruthRewards instead of the reward model's predicted rewards.

**Comment by jualat** [comment]
> > > Looks good, but we need to adjust where the groundTruthRewards are stored. With this, the rewards are updated in the buffer and used for comparison in the teacher. However, for accurate comparisons in the teacher, we should still use groundTruthRewards instead of the reward model's predicted rewards.
> > 
> > 
> > Maybe we should add a new tensor to the replay buffer called groundTruthRewards which is shaped equally like the "normal" rewards tensor from the parent (StableBaselines replay buffer). With this approach, we could also get rid of the second replay buffer you added in #19, right?
> 
> Yes with this approach we could remove the second replay buffer and could start training the reward model right after the exploration has finished

Sounds like a good approach to me, will implement this :+1:

File: poetry.lock
Comment (bkuen) [reply poetry.lock:1]: This approach seems fine to me. Let's do it. I think, this branch is a good way to start with it, as #17 probably doesn't add any further dependencies.

**Comment by jualat** [comment]
Thanks for the comment, with this change I can actually see a performance increase now, even with the old and small reward model.
image

File: framework/replay_buffer.py [hunk 1]
```diff
@@ -84,6 +99,27 @@ def sample_trajectories(self, env: Optional[VecNormalize] = None):
...
+        data = (
+            self._normalize_obs(self.observations[batch_inds, env_indices, :], env),
+            self.actions[batch_inds, env_indices, :],
+            next_obs,
+            # Only use dones that are not due to timeouts
+            # deactivated by default (timeouts is initialized as an array of False)
+            (self.dones[batch_inds, env_indices] * (1 - self.timeouts[batch_inds, env_indices])).reshape(-1, 1),
+            self._normalize_reward(self.ground_truth_rewards[batch_inds, env_indices].reshape(-1, 1), env),
+        )
+        return ReplayBufferSamples(*tuple(map(self.to_torch, data)))
```
Comment (Wizzardexe) [inline framework/replay_buffer.py:121]: We should consider replacing ReplayBufferSamples() with a custom implementation, such as CustomReplayBufferSamples(NamedTuple), because ReplayBufferSamples() limits us to six positional arguments, which does not meet our requirement for seven arguments. [This is how I suggest we could implement it](https://github.com/jualat/CleanRLHF/tree/18-relabeling-added-CustomReplayBufferSamples)

File: framework/replay_buffer.py [hunk 2]
```diff
@@ -39,6 +41,8 @@ class ReplayBuffer(SB3ReplayBuffer):
         https://github.com/DLR-RM/stable-baselines3/issues/284
     """
 
+    ground_truth_rewards: np.ndarray
```
[the remark this answers is not in the document]
Comment (jualat) [reply framework/replay_buffer.py:44]: true :+1:

File: framework/replay_buffer.py (hunk 1 above)
Comment (jualat) [reply framework/replay_buffer.py:121]: Yes, you're right, implementation should follow in some minutes :+1:

File: framework/replay_buffer.py [hunk 3]
```diff
@@ -82,6 +99,27 @@ def sample_trajectories(self, env: Optional[VecNormalize] = None):
...
+            next_obs = self._normalize_obs(self.next_observations[batch_inds, env_indices, :], env)
+
+        data = (
+            self._normalize_obs(self.observations[batch_inds, env_indices, :], env),
+            self.actions[batch_inds, env_indices, :],
+            next_obs,
+            # Only use dones that are not due to timeouts
+            # deactivated by default (timeouts is initialized as an array of False)
+            (self.dones[batch_inds, env_indices] * (1 - self.timeouts[batch_inds, env_indices])).reshape(-1, 1),
+            self._normalize_reward(self.ground_truth_rewards[batch_inds, env_indices].reshape(-1, 1), env),
```
[the remark this answers is not in the document]
Comment (jualat) [reply framework/replay_buffer.py:119]: right, forgot about this

File: framework/sac_rlhf.py [hunk 4]
```diff
@@ -349,7 +349,7 @@ def get_action(self, x):
             intrinsic_reward = knn_estimator.compute_intrinsic_rewards(next_obs)
             knn_estimator.update_states(next_obs)
 
-            rb_exp.add(obs, real_next_obs, actions, intrinsic_reward, terminations, infos)
+            rb_exp.add(obs, real_next_obs, actions, intrinsic_reward, ground_truth_reward, terminations, infos)
```
Comment (jualat) [inline framework/sac_rlhf.py:352]: Didn't had a deeper look into the unsuperviced pretraining part yet, @Wizzardexe can you have a look if this is right? Shouldn't have any impact as far as I see.

Comment (Wizzardexe) [reply framework/sac_rlhf.py:352]: I would remove the extra replay buffer and just use rb.add here.

File: framework/sac_rlhf.py [hunk 5]
```diff
@@ -447,7 +447,9 @@ def get_action(self, x):
         for idx, trunc in enumerate(truncations):
             if trunc:
                 real_next_obs[idx] = infos["final_observation"][idx]
-        rb.add(obs, real_next_obs, actions, groundTruthRewards, terminations, infos)
+
+        rewards = reward_net.predict_reward(obs, actions)
```
Comment (Wizzardexe) [inline framework/sac_rlhf.py:451]: You need to use rewards = rewards.detach().cpu().numpy() here.

Comment (jualat) [reply framework/sac_rlhf.py:451]: fair :+1:

File: framework/sac_rlhf.py (hunk 4 above)
Comment (jualat) [reply framework/sac_rlhf.py:352]: done :+1:

**Review by Wizzardexe — APPROVED** [review]
Looks good to me. One thing we can change now is that we don't need to differentiate between exploration and training in the `train_q_network()` function, but either you can change this or we can change this in another pull request.

**Review by bkuen — APPROVED** [review]
